File: dotnet/harness/Relay.Testing/UnfurlHarness.cs
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc.Testing;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;

namespace Relay.Testing;

/// <summary>
/// Outbound-HTTP unfurl harness (spec/04-dependencies.md §7): a REAL local stub HTTP
/// server with fault injection — NOT an in-process mock of the client class. The timeout,
/// the socket and the status codes are real, which is the whole point: resilience lives
/// in the real client config + socket behavior, invisible to a mock. <c>Seed</c> = program
/// the route (200+title / delay &gt; timeout / 500); <c>Assert</c> = received-request count
/// (the circuit-breaker proof); <c>Reset</c> = clear the programmed route + counter.
/// </summary>
public sealed class UnfurlHarness<TProgram> : IDependencyHarness<TProgram>
    where TProgram : class
{
    private readonly Lock _lock = new();
    private WebApplication? _stub;
    private string? _baseUrl;
    private Behavior _behavior = Behavior.Ok("Example");
    private int _requestCount;

    public string BaseUrl => _baseUrl ?? throw new InvalidOperationException("UnfurlHarness is not started.");

    public int RequestCount
    {
        get { lock (_lock) { return _requestCount; } }
    }

    public void ConfigureWebHostBuilder(IWebHostBuilder builder) =>
        builder.UseSetting("Unfurl:BaseUrl", BaseUrl);

    public async Task Start(WebApplicationFactory<TProgram> factory, CancellationToken cancellationToken)
    {
        var builder = WebApplication.CreateBuilder();
        builder.WebHost.UseUrls("http://127.0.0.1:0");
        builder.Logging.ClearProviders();
        _stub = builder.Build();

        _stub.MapGet("/unfurl", async (HttpContext context) =>
        {
            Behavior behavior;
            lock (_lock)
            {
                _requestCount++;
                behavior = _behavior;
            }

            await behavior.Apply(context);
        });

        await _stub.StartAsync(cancellationToken);
        _baseUrl = _stub.Urls.First();
    }

    public async Task Stop(CancellationToken cancellationToken)
    {
        if (_stub is not null)
        {
            await _stub.StopAsync(cancellationToken);
            await _stub.DisposeAsync();
        }
    }

    public void ProgramOk(string title) => SetBehavior(Behavior.Ok(title));

    public void ProgramDelay(TimeSpan delay, string title = "Example") => SetBehavior(Behavior.Delayed(delay, title));

    public void ProgramServerError() => SetBehavior(Behavior.ServerError());

    public void Reset()
    {
        lock (_lock)
        {
            _behavior = Behavior.Ok("Example");
            _requestCount = 0;
        }
    }

    private void SetBehavior(Behavior behavior)
    {
        lock (_lock)
        {
            _behavior = behavior;
        }
    }

    private sealed record Behavior(int Status, string? Title, TimeSpan Delay)
    {
        public static Behavior Ok(string title) => new(200, title, TimeSpan.Zero);

        public static Behavior Delayed(TimeSpan delay, string title) => new(200, title, delay);

        public static Behavior ServerError() => new(500, null, TimeSpan.Zero);

        public async Task Apply(HttpContext context)
        {
            if (Delay > TimeSpan.Zero)
            {
                await Task.Delay(Delay, context.RequestAborted);
            }

            context.Response.StatusCode = Status;
            if (Status == 200)
            {
                await context.Response.WriteAsJsonAsync(new { title = Title });
            }
        }
    }
}

File: java/src/test/java/dev/drim/relay/harness/UnfurlHarness.java
package dev.drim.relay.harness;

import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * The outbound-HTTP harness: a REAL local stub server (not an in-process client mock — the timeout,
 * the socket, and the status codes must be real). Seed = program the route (200+title / delay &gt;
 * timeout / 500 / connection reset); Assert = received-request count (circuit-breaker proof); Reset
 * = clear route + counter. Mirrors go/harness/unfurl.go.
 */
public final class UnfurlHarness implements DependencyHarness {
  private enum Mode {
    OK,
    DELAY,
    SERVER_ERROR,
    RESET
  }

  private final Object lock = new Object();
  private final AtomicInteger requests = new AtomicInteger(0);

  private HttpServer server;
  private Mode mode = Mode.OK;
  private String title = "Example";
  private long delayMs = 0;

  @Override
  public void start() {
    try {
      server = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);
      server.createContext("/unfurl", this::handle);
      server.start();
    } catch (IOException e) {
      throw new IllegalStateException("unfurl stub start failed", e);
    }
    mode = Mode.OK;
    title = "Example";
  }

  public String baseUrl() {
    return "http://127.0.0.1:" + server.getAddress().getPort();
  }

  /** The number of /unfurl requests the stub received (breaker proof). */
  public int requestCount() {
    return requests.get();
  }

  @Override
  public void reset() {
    synchronized (lock) {
      mode = Mode.OK;
      title = "Example";
      delayMs = 0;
    }
    requests.set(0);
  }

  @Override
  public void stop() {
    if (server != null) {
      server.stop(0);
    }
  }

  /** Programs a 200 response with the given title. */
  public void programOk(String title) {
    synchronized (lock) {
      this.mode = Mode.OK;
      this.title = title;
    }
  }

  /** Programs a response slower than the 800 ms client timeout. */
  public void programDelay(long delayMs) {
    synchronized (lock) {
      this.mode = Mode.DELAY;
      this.delayMs = delayMs;
    }
  }

  /** Programs a 500 response. */
  public void programServerError() {
    synchronized (lock) {
      this.mode = Mode.SERVER_ERROR;
    }
  }

  /** Programs an abrupt connection close (no HTTP response). */
  public void programConnectionReset() {
    synchronized (lock) {
      this.mode = Mode.RESET;
    }
  }

  private void handle(HttpExchange exchange) throws IOException {
    requests.incrementAndGet();
    Mode m;
    String t;
    long d;
    synchronized (lock) {
      m = mode;
      t = title;
      d = delayMs;
    }
    switch (m) {
      case DELAY -> {
        sleep(d);
        respondJson(exchange, t);
      }
      case SERVER_ERROR -> {
        exchange.sendResponseHeaders(500, -1);
        exchange.close();
      }
      case RESET -> exchange.close();
      default -> respondJson(exchange, t);
    }
  }

  private static void respondJson(HttpExchange exchange, String title) throws IOException {
    byte[] body =
        ("{\"title\":\"" + title.replace("\"", "\\\"") + "\"}").getBytes(StandardCharsets.UTF_8);
    exchange.getResponseHeaders().set("Content-Type", "application/json");
    exchange.sendResponseHeaders(200, body.length);
    exchange.getResponseBody().write(body);
    exchange.close();
  }

  private static void sleep(long ms) {
    try {
      Thread.sleep(ms);
    } catch (InterruptedException e) {
      Thread.currentThread().interrupt();
    }
  }
}
